servo-media/src/backends/gstreamer/audio_decoder.rs

```diff
@@ -0,0 +1,195 @@
+            None => return callbacks.error(),
+        };
+
+        let decodebin = match gst::ElementFactory::make("decodebin", None) {
+            Some(decodebin) => decodebin,
+            None => return callbacks.error(),
+        };
+
+        // decodebin uses something called a "sometimes-pad", which is basically
+        // a pad that will show up when a certain condition is met,
+        // in decodebins case that is media being decoded
+        if pipeline.add_many(&[&appsrc, &decodebin]).is_err() {
+            return callbacks.error();
+        }
+
+        if gst::Element::link_many(&[&appsrc, &decodebin]).is_err() {
+            return callbacks.error();
+        }
+
+        let appsrc = match appsrc.downcast::<AppSrc>() {
+            Ok(appsrc) => appsrc,
+            Err(_) => {
+                return callbacks.error();
+            }
+        };
+
+        let pipeline_ = pipeline.clone();
+        let callbacks_ = callbacks.clone();
+        decodebin.connect_pad_added(move |_, src_pad| {
+            // We only want a sink per audio stream.
+            if src_pad.is_linked() {
+                return;
+            }
+
+            let pipeline = &pipeline_;
+            let callbacks = &callbacks_;
+
+            let (is_audio, _) = {
+                let media_type = src_pad.get_current_caps().and_then(|caps| {
+                    caps.get_structure(0).map(|s| {
+                        let name = s.get_name();
+                        (name.starts_with("audio/"), name.starts_with("video/"))
+                    })
+                });
+
+                match media_type {
+                    None => {
+                        eprintln!("Failed to get media type from pad {}", src_pad.get_name());
+                        return callbacks.error();
+                    }
+                    Some(media_type) => media_type,
+                }
+            };
+
+            if !is_audio {
+                return callbacks.error();
+            }
+
+            let insert_sink = || -> Result<(), ()> {
+                let queue = gst::ElementFactory::make("queue", None).ok_or(())?;
+                let convert = gst::ElementFactory::make("audioconvert", None).ok_or(())?;
+                let resample = gst::ElementFactory::make("audioresample", None).ok_or(())?;
+                let sink = gst::ElementFactory::make("appsink", None).ok_or(())?;
+                let appsink = sink.clone().dynamic_cast::<AppSink>().map_err(|_| ())?;
+
+                // XXX Issue #18 channels support.
+                let audio_info = gst_audio::AudioInfo::new(gst_audio::AUDIO_FORMAT_F32, 44100, 1)
+                    .build()
+                    .ok_or(())?;
+                appsink.set_caps(&audio_info.to_caps().unwrap());
+
+                let pipeline_ = pipeline.clone();
+                let pipeline__ = pipeline.clone();
+                let callbacks_ = callbacks.clone();
+                let callbacks__ = callbacks.clone();
```

Comment:
not too happy about this but this seems to make sense 😄

Reply:
Yeah, I am not too happy either, but I couldn't think of a better approach :\